### PR TITLE
Static tokens for kubeconfigs + rotation

### LINKED ISSUE #105: Admin Password Reset
# Story
- As user I want to reset the password, so that I can change it if it leaked.

# Motivation
Users sometimes/accidentally leak their admin password.

# Acceptance Criteria
First we must answer the question whether we like to just reset it or also set it, i.e.
- [ ] User can request admin password regeneration through the shoot cluster resource

or
- [ ] User can set the admin password through the shoot cluster resource

# Definition of Done
- [ ] Knowledge is distributed: Have you spread your knowledge in pair programming/code review?
- [ ] Unit Tests are provided: Have you written automated unit tests or added manual test tickets?
- [ ] Integration Tests are provided: Have you written automated integration tests or added manual test tickets?
- [ ] Minimum API exposure: If you have added public API, was it really necessary/is it minimal?
- [ ] Operations guide: Have you updated the operations guide about ops-relevant changes?
- [ ] User documentation: Have you informed the documentation developers about user-relevant changes?

### PR DESCRIPTION
**Pull request by rfranzke**
**What this PR does / why we need it**:
The kubeconfig handed out to end-users does no longer contain a client certificate. The motivation for this is that leaked client certificate cannot be revoked without exchanging the cluster's CA (which is a non-trivial operation). Instead, it will contain a static token that is used for authentication against the shoot's kube-apiserver.
It is possible to trigger a rotation of this static token and the basic authentication password by annotating the `Shoot` with `shoot.garden.sapcloud.io/operation=rotate-kubeconfig-credentials`.
:warning: If this annotation is set to an cluster created with a prior Gardener version then the newly generated kubeconfig will also only contain a static token, however, the previously issued client certificate will still remain valid.

**Which issue(s) are fixed by this PR**:
fixes #105 

**Special notes for your reviewer**:
* /cc @holgerkoser @grolu @petersutter @vasu1124 @vlerenc @ccwienk @ThormaehlenFred 
* FYI: After this one is merged I'll create a follow-up PR that will allow to disable basic authentication for a shoot (default will still be `true`).
* Along the way I contexified the `DeploySecrets` function.

**Release note**:

```action user
The kubeconfigs for newly created shoot clusters do no longer contain client certificates. Instead, they contain a token that can be rotated on your demand. When annotation the shoot with `shoot.garden.sapcloud.io/operation=rotate-kubeconfig-credentials` this token will be rotated.
```
```noteworthy user
You may now rotate the basic authentication password of your shoot cluster by annotating the `Shoot` resource with `shoot.garden.sapcloud.io/operation=rotate-kubeconfig-credentials`. :warning: Please be aware that if this cluster was created before Gardener version `0.28.0` then your new kubeconfig will no longer contain a client certificate (however, it the previously issued one will remain valid).
```

### CODE REVIEW

File: pkg/operation/botanist/secrets.go [hunk 1]
```diff
@@ -650,8 +650,32 @@ func (b *Botanist) generateWantedSecrets(basicAuthAPIServer *secrets.BasicAuth,
...
-func (b *Botanist) DeploySecrets() error {
-	existingSecretsMap, err := b.fetchExistingSecrets()
+func (b *Botanist) DeploySecrets(ctx context.Context) error {
+	// If the rotate-kubeconfig operation annotation is set then we delete the existing kubecfg and basic-auth
+	// secrets. This will trigger the regeneration, incorporating new credentials. After successful deletion of all
+	// old secrets we remove the operation annotation.
+	if kutil.HasMetaDataAnnotation(b.Shoot.Info, common.ShootOperation, common.ShootOperationRotateKubeconfigCredentials) {
+		b.Logger.Infof("Rotating kubeconfig credentials")
+
+		if err := b.K8sSeedClient.Client().Delete(ctx, &corev1.Secret{ObjectMeta: metav1.ObjectMeta{Name: common.StaticTokenSecretName, Namespace: b.Shoot.SeedNamespace}}); client.IgnoreNotFound(err) != nil {
```
Comment (mvladev) [inline pkg/operation/botanist/secrets.go]: Those deletions are more or less the same - use `for` loop?

File: pkg/utils/secrets/static_token.go [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+func (s *StaticTokenSecretConfig) Generate() (Interface, error) {
+	return s.GenerateStaticToken()
+}
+
+// GenerateStaticToken computes a random token of length 64.
+func (s *StaticTokenSecretConfig) GenerateStaticToken() (*StaticToken, error) {
+	tokens := make([]Token, 0, len(s.Tokens))
+
+	for _, tokenConfig := range s.Tokens {
+		token, err := utils.GenerateRandomString(64)
```
Comment (mvladev) [inline pkg/utils/secrets/static_token.go]: Can we increase the size to `128`?

File: pkg/operation/botanist/secrets.go (hunk 1 above)
Comment (rfranzke) [reply pkg/operation/botanist/secrets.go]: OK, will do 👍

File: pkg/utils/secrets/static_token.go (hunk 2 above)
Comment (rfranzke) [reply pkg/utils/secrets/static_token.go]: Sure, we can. Will do.

File: pkg/registry/garden/shoot/strategy.go [hunk 3]
```diff
@@ -97,9 +97,16 @@ func mustIncreaseGeneration(oldShoot, newShoot *garden.Shoot) bool {
...
 		default:
-			// The shoot state is not failed and the reconcile annotation is set.
-			if val, ok := newShoot.Annotations[common.ShootOperation]; ok && val == common.ShootOperationReconcile {
-				mustIncrease = true
+			// The shoot state is not failed and the reconcile or rotate-credentials annotation is set.
+			if val, ok := newShoot.Annotations[common.ShootOperation]; ok {
+				if val == common.ShootOperationReconcile {
+					mustIncrease = true
+				}
+				if val == common.ShootOperationRotateKubeconfigCredentials {
```
Comment (mvladev) [inline pkg/registry/garden/shoot/strategy.go:128]: Can you add documentation for this?

We have lots of annotations and they are very undocumented.

Comment (rfranzke) [reply pkg/registry/garden/shoot/strategy.go:128]: Done

File: pkg/utils/secrets/secrets.go [hunk 4]
```diff
@@ -50,7 +52,7 @@ func (s *Secrets) Deploy(
 
 	// Generate cluster secrets
 	secretConfigs := s.SecretConfigsFunc(cas, namespace)
-	clusterSecrets, err := GenerateClusterSecrets(gcs, existingSecrets, secretConfigs, namespace)
+	clusterSecrets, err := GenerateClusterSecrets(context.TODO(), gcs, existingSecrets, secretConfigs, namespace)
```
Comment (timuthy) [inline pkg/utils/secrets/secrets.go]: Can you let the `context` be passed to `func (s *Secrets) Deploy` (line 35) and use it here?

Comment (rfranzke) [reply pkg/utils/secrets/secrets.go]: done

**Review by timuthy — APPROVED** [review]
/lgtm - nice!

File: docs/usage/shoot_operations.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
+# Trigger shoot operations
```
Comment (ialidzhikov) [inline docs/usage/shoot_operations.md:1]: Nice docs.
I will remove the https://github.com/gardener/documentation/blob/master/website/documentation/050-tutorials/content/howto/maintain-shoot/_index.md#shoot-annotations and link this file as separate page.
Is it okay for you?

Comment (rfranzke) [reply docs/usage/shoot_operations.md:1]: Yes, sure, thanks! Didn't know that this already exists.

**Pull request merged**